File: js/complete/no039.js
#!/usr/bin/env node

/* For p, we seek k, m, n > 0 such that n < m and 2*k*m*(m + n) = p */

var fns = require('../functions.js'),
    operator = require('../operator.js'),
    timer = require('../timer.js');

function allTriples(p, factorsHash) {
    if (typeof factorsHash == 'undefined') {
        factorsHash = {};
    }

    if (p % 2 || p < 2) {
        return [];
    }

    var choicesK;
    // We know p % 2 == 0
    if (p / 2 in factorsHash) {
        choicesK = factorsHash[p / 2];
    } else {
        choicesK = fns.allFactors(p / 2, factorsHash)[p / 2];
    }

    var result = [], choicesM, j, m, n;
    for (var i = 0, k; k = choicesK[i]; i++) {
        if (p / (2 * k) in factorsHash) {
            choicesM = factorsHash[p / (2 * k)];
        } else {
            choicesM = fns.allFactors(p / (2 * k), factorsHash)[p / (2 * k)];
        }

        // 2 * k * m * (m + n) == p
        for (j = 0; m = choicesM[j]; j++) {
            n = p / (2 * k * m) - m;
            if (n > 0 && m > n) {
                result.push([k, m, n]);
            }
        }
    }

    return result;
};

function convertToTriangle(triple) {
    var k = triple[0], m = triple[1], n = triple[2],
        a = k * (m * m - n * n),
        b = k * (2 * m * n),
        c = k * (m * m + n * n);
    return operator.uniqSorted([a, b, c]);
};

function allTriangles(p, factorsHash) {
    if (typeof factorsHash == 'undefined') {
        factorsHash = {};
    }

    var triples = allTriples(p, factorsHash);
    function stringMap(triangle) {
        return triangle.join(', ');
    };

    function undoStringMap(str) {
        return str.split(', ').map(Number);
    };
    return operator.uniqSorted(triples.map(convertToTriangle).map(stringMap), false).map(undoStringMap);
};

function allTrianglesUpToN(n) {
    var factorsHash = fns.allFactors(n), result = {};
    for (var p = 2; p <= n; p += 2) {
        result[p] = allTriangles(p, factorsHash);
    }
    return result;
};

exports.main = function(verbose) {
    if (typeof verbose == 'undefined') {
        verbose = false;
    }
    var allTri = allTrianglesUpToN(1000), lengths = {}, maxVal = -1, maxKeys = [], key;
    for (key in allTri) {
        currLength = allTri[key].length;
        if (currLength > maxVal) {
            maxKeys = [key];
            maxVal = currLength;
        } else if (currLength == maxVal) {
            maxKeys.push(key);
        }
    }

    if (maxKeys.length != 1) {
        return; // Keys are not unique
    }
    return maxKeys[0];
};

if (require.main === module) {
    timer.timer(39, exports.main);
}

File: python/complete/no039.py
#!/usr/bin/env python

# For p, we seek k, m, n > 0 such that n < m and 2*k*m*(m + n) = p

from python.decorators import euler_timer
from python.functions import all_factors


def all_triples(p, factors_hash=None):
    if factors_hash is None:
        factors_hash = {}

    if p % 2 == 1 or p < 2 or not isinstance(p, int):
        return []

    if p / 2 in factors_hash:
        choices_k = factors_hash[p / 2]
    else:
        choices_k = all_factors(p / 2, factors_hash)[p / 2]

    result = []
    for k in choices_k:
        if p / (2 * k) in factors_hash:
            choices_m = factors_hash[p / (2 * k)]
        else:
            choices_m = all_factors(p / (2 * k), factors_hash)[p / (2 * k)]

        # 2*k*m*(m + n) = p
        for m in choices_m:
            n = p / (2 * k * m) - m
            if n > 0 and m > n:
                result.append((k, m, n))

    return result


def convert_to_triangle(triple):
    k, m, n = triple
    a = k * (m ** 2 - n ** 2)
    b = k * (2 * m * n)
    c = k * (m ** 2 + n ** 2)
    return tuple(sorted((a, b, c)))


def all_triangles(p, factors_hash=None):
    triples = all_triples(p, factors_hash)
    return list(set(convert_to_triangle(triple) for triple in triples))


def all_triangles_up_to_n(n):
    factors_hash = all_factors(n)
    result = {}
    for p in range(2, n + 1, 2):
        result[p] = all_triangles(p, factors_hash)
    return result


def main(verbose=False):
    all_tri = all_triangles_up_to_n(1000)
    lengths = {}
    max_val = -1
    max_keys = []
    for key, value in all_tri.iteritems():
        curr_length = len(value)
        if curr_length > max_val:
            max_keys = [key]
            max_val = curr_length
        elif curr_length == max_val:
            max_keys.append(key)

    if len(max_keys) != 1:
        raise("Keys are not unique")

    return max_keys[0]

if __name__ == '__main__':
    print euler_timer(39)(main)(verbose=True)
